aten/src/ATen/core/ITensorList.h

```diff
@@ -0,0 +1,367 @@
+#pragma once
+
+#include <ATen/core/List.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Exception.h>
+
+#include <initializer_list>
+#include <iterator>
+#include <type_traits>
+
+namespace at {
+class Tensor;
+}
+
+namespace c10 {
+class ITensorList;
+class ITensorListIterator;
+
+/*
+ * Adding a new tag
+ * ================
+
+ * -> Short Version
+ *
+ * If you are lucky, the only necessary things for introducing
+ * a new tag are:
+ *
+ *     - Add a line to `TORCH_ITENSORLIST_FORALL_TAGS` such as:
+ *                 `_(TagName, ##__VA_ARGS__)`
+ *
+ *     - Specialize `ITensorListTagImpl` to your tag, implementing
+ *       the required methods
+ *
+ *     - Implement constructors for the type corresponding to your
+ *       tag on both `ITensorList(Iterator)` classes
+ *
+ * -> Long Version
+ *
+ * That may not work if your container is "more different" than
+ * the ones implemented. Specifically, maybe, some of the functions
+ * implemented with `TORCH_ITENSORLIST_UNWRAP` may not work for it.
+ * In those cases, you should do something like `get` and
+ * `iterator_get` functions.
+ *
+ * Say the implementation for `begin` is different. Then, you should:
+ *
+ *     - Do everything done in the "Short Version" for introducing
+ *       the new tag
+ *
+ *     - Implement a new function `begin` in all implementation
+ *       classes (for all tag types)
+ *
+ *     - Call it directly inside a `TORCH_ITENSORLIST_UNWRAP`
+ */
+
+// Applies arbitrary macros to each `ITensorListTag`.
+#define TORCH_ITENSORLIST_FORALL_TAGS(_, ...) \
+  _(Unboxed, ##__VA_ARGS__)                   \
+  _(Boxed, ##__VA_ARGS__)
```

Comment:
Hmm... do you think that the macros here reduce boilerplate enough to warrant their extra complexity? My worry is that there's enough uncertainty about what "the next ITensorList impl" will look like that, like you mentioned above, adding it in wouldn't just be a simple "add your name to this line" change. It'd require reading through and understanding all of the macro logic here, which is pretty non-trivial.

I wonder how much more verbose all the code here would be without macros.

Reply:
> there's enough uncertainty about what "the next ITensorList impl" will look like that, like you mentioned above

I think I have an idea for how to extend it for optional tensors. Most of the `ITensorList` implementation would stay the same (although, I think I would go back and call it `IList`), including the macros.

> do you think that the macros here reduce boilerplate enough to warrant their extra complexity?

Hmmm. That's a fair point.
But, I do believe the macros do more good than bad. I decided to use them since most of the implementations for both containers would be the same. Otherwise, I think the code would be full of switches and copy-pasted code (e.g. replace every `TORCH_ITENSORLIST_UNWRAP` by a switch and a `TORCH_CHECK` in the end) . I'd even say that once you know what one understands what `TORCH_ITENSORLIST_UNWRAP` does, the code gets easier to read/modify.